File: backend/src/repositories/requestRepository.ts
import { GetCommand, PutCommand, QueryCommand, UpdateCommand } from "@aws-sdk/lib-dynamodb";
import { dynamo } from "../db/dynamo.js";

const tableName = process.env.DYNAMODB_REQUESTS_TABLE;

if (!tableName) {
  throw new Error("Missing DYNAMODB_REQUESTS_TABLE");
}

export type RequestRecord = {
  request_id: string;
  requested_by: string;
  requested_by_sub: string;
  provider: string;
  template_id: string;
  parameters: Record<string, unknown>;
  status: string;
  branch_name: string;
  pr_url: string;
  pr_number: number;
  created_at: string;
  updated_at: string;
};

export async function createRequest(record: RequestRecord): Promise<void> {
  await dynamo.send(
    new PutCommand({
      TableName: tableName,
      Item: record,
    })
  );
}

export async function getRequestById(
  requestId: string
): Promise<RequestRecord | undefined> {
  const result = await dynamo.send(
    new GetCommand({
      TableName: tableName,
      Key: {
        request_id: requestId,
      },
    })
  );

  return result.Item as RequestRecord | undefined;
}

export async function listRequestsByUser(
  requestedBySub: string
): Promise<RequestRecord[]> {
  const result = await dynamo.send(
    new QueryCommand({
      TableName: tableName,
      IndexName: "requested_by_sub-created_at-index",
      KeyConditionExpression: "requested_by_sub = :requested_by_sub",
      ExpressionAttributeValues: {
        ":requested_by_sub": requestedBySub,
      },
      ScanIndexForward: false,
    })
  );

  return (result.Items ?? []) as RequestRecord[];
}

export async function updateRequestStatus(
  requestId: string,
  status: string
): Promise<void> {
  await dynamo.send(
    new UpdateCommand({
      TableName: tableName,
      Key: {
        request_id: requestId,
      },
      UpdateExpression: "SET #status = :status, updated_at = :updated_at",
      ExpressionAttributeNames: {
        "#status": "status",
      },
      ExpressionAttributeValues: {
        ":status": status,
        ":updated_at": new Date().toISOString(),
      },
    })
  );
}

File: backend2/app/repositories/request_repository.py
from app.config import settings
from app.db.dynamo import dynamodb

if not settings.DYNAMODB_REQUESTS_TABLE:
    raise RuntimeError("Missing DYNAMODB_REQUESTS_TABLE")

table = dynamodb.Table(settings.DYNAMODB_REQUESTS_TABLE)

def create_request(record: dict) -> None:
    table.put_item(Item=record)

def get_request_by_id(request_id: str) -> dict | None:
    response = table.get_item(Key={"request_id": request_id})
    return response.get("Item")

def list_requests_by_user(requested_by_sub: str) -> list[dict]:
    response = table.query(
        IndexName="requested_by_sub-created_at-index",
        KeyConditionExpression="requested_by_sub = :requested_by_sub",
        ExpressionAttributeValues={
            ":requested_by_sub": requested_by_sub
        },
        ScanIndexForward=False
    )
    return response.get("Items", [])

def update_request_status(request_id: str, status: str) -> None:
    from datetime import datetime, timezone

    table.update_item(
        Key={"request_id": request_id},
        UpdateExpression="SET #status = :status, updated_at = :updated_at",
        ExpressionAttributeNames={"#status": "status"},
        ExpressionAttributeValues={
            ":status": status,
            ":updated_at": datetime.now(timezone.utc).isoformat()
        },
    )